### PR TITLE
Use formik for password change form

### LINKED ISSUE #414: Use Formik for password change forms

### PR DESCRIPTION
**Pull request by yedhink**
Fixes #414 

@vinay0x _a
Can you review and merge this?

- Working with `Formik` for the first time
- Do let me know, if you find some anti-patterns

[Edit]
- Made the server side code to update  password api based.
- Modified the test case associated with it.

 Thanks

### CODE REVIEW

File: app/controllers/registrations_controller.rb [hunk 1]
```diff
@@ -5,15 +5,12 @@ class RegistrationsController < Devise::RegistrationsController
...
   def update_password
-    if update_resource(resource, password_update_params)
-      if is_flashing_format?
-        set_flash_message :notice, :updated
-        bypass_sign_in(resource)
-        respond_with(resource, location: after_update_path_for(resource))
-      end
+    if resource.update_with_password(password_update_params)
+      bypass_sign_in resource, scope: :user
+      render json: { notice: 'Successfully updated password' }, status: :ok
```
Comment (vinay0x) [inline app/controllers/registrations_controller.rb:10]:
```suggestion
      render json: { notice: 'Password has been successfully updated' }, status: :ok
```

File: app/controllers/registrations_controller.rb [hunk 2]
```diff
@@ -5,15 +5,12 @@ class RegistrationsController < Devise::RegistrationsController
...
     else
       clean_up_passwords resource
-      render json: { error: "Couldnt' update the password! Please try again." }, status: 422
+      render json: { error: "Couldn't update the password! Make sure you entered your current password correctly." }, status: 422
```
Comment (vinay0x) [inline app/controllers/registrations_controller.rb:13]: We can't say this. We don't know for sure if it's the reason for the password not being updated properly.
```suggestion
      render json: { error: "Couldn't update the password! Please try again." }, status: 422
```

File: app/javascript/src/apis/registrations.js [hunk 3]
```diff
@@ -0,0 +1,9 @@
+import axios from "axios";
+
+const updatePassword = payload => axios.put("my/password/update", payload);
+
+const registrationsApi = {
+  updatePassword,
+};
+
+export default registrationsApi;
```
Comment (vinay0x) [inline app/javascript/src/apis/registrations.js:9]:
```suggestion
export default {
  updatePassword,
};
```

File: app/javascript/src/components/Dashboard/Account/Passwords/Edit.jsx [hunk 4]
```diff
@@ -1,9 +1,41 @@
...
+import { Button } from "neetoui";
+
+import registrationsApi from "apis/registrations";
 
 const Edit = () => {
-  const [currentPassword, setCurrentPassword] = useState("");
-  const [password, setPassword] = useState("");
-  const [passwordConfirmation, setPasswordConfirmation] = useState("");
+  const validationSchema = Yup.object({
+    currentPassword: Yup.string().required("Current Password is required"),
```
Comment (vinay0x) [inline app/javascript/src/components/Dashboard/Account/Passwords/Edit.jsx:11]: Use sentence case and only capitalize words where it makes sense to: names, places etc.
```suggestion
    currentPassword: Yup.string().required("Current password is required"),
```

File: app/javascript/src/components/Dashboard/Account/Passwords/Edit.jsx [hunk 5]
```diff
@@ -1,9 +1,41 @@
...
+    password: Yup.string().required("New Password is required"),
```
Comment (vinay0x) [inline app/javascript/src/components/Dashboard/Account/Passwords/Edit.jsx:12]:
```suggestion
    password: Yup.string().required("New password is required"),
```

File: app/javascript/src/components/Dashboard/Account/Passwords/Edit.jsx [hunk 6]
```diff
@@ -1,9 +1,41 @@
...
+    passwordConfirmation: Yup.string()
+      .required("Confirmation Password is required")
```
Comment (vinay0x) [inline app/javascript/src/components/Dashboard/Account/Passwords/Edit.jsx:14]:
```suggestion
      .required("Password confirmation is required")
```

File: app/javascript/src/components/Dashboard/Account/Passwords/Edit.jsx [hunk 7]
```diff
@@ -1,9 +1,41 @@
...
+      .oneOf([Yup.ref("password"), null], "Passwords must match"),
+  });
+
+  const formikInputAttrs = {
+    type: "password",
+    "aria-required": "true",
+    placeholder: "******",
+    className:
+      "w-full px-3 py-2 text-gray-800 transition duration-200 ease-in-out rounded-md form-control focus:text-black hover:border-gray-600 focus:border-gray-600 focus:outline-none",
```
Comment (vinay0x) [inline app/javascript/src/components/Dashboard/Account/Passwords/Edit.jsx:23]: Why are all these classes required? Input component is already styled.

File: app/javascript/src/components/Dashboard/Account/Passwords/Edit.jsx [hunk 8]
```diff
@@ -1,9 +1,41 @@
...
+
+  const handleSubmit = async data => {
+    try {
+      await registrationsApi.updatePassword({
+        user: {
+          current_password: data.currentPassword,
+          password: data.password,
+          password_confirmation: data.passwordConfirmation,
+        },
+      });
```
Comment (vinay0x) [inline app/javascript/src/components/Dashboard/Account/Passwords/Edit.jsx:34]: Show success toast message.

File: app/javascript/src/components/Dashboard/Account/Passwords/Edit.jsx [hunk 9]
```diff
@@ -12,85 +44,44 @@ const Edit = () => {
...
+                {...formikInputAttrs}
+                name="passwordConfirmation"
+                id="password_confirmation"
+                label="Confirm Password"
+              />
+              <div className="flex justify-center items-center w-full p-2">
+                <Button
+                  name="submit"
+                  type="submit"
+                  className="flex justify-center items-center w-full text-base font-semibold text-white transition duration-200 ease-in-out bg-teal-600 rounded-md cursor-pointer hover:opacity-75"
```
Comment (vinay0x) [inline app/javascript/src/components/Dashboard/Account/Passwords/Edit.jsx:79]: Again, why is this required?

**Review by vinay0x — CHANGES_REQUESTED** [review]
@yedhink _a

File: app/javascript/src/apis/registrations.js (hunk 3 above)
Comment (yedhink) [reply app/javascript/src/apis/registrations.js:9]: @vinay0x 

This was intentional. It helps a lot with the intellisense. We don't have to manually import from the apis if we write like this. We could just type in `registrationsApi` and vscode will auto import for us.

This widely helps with faster writing of code with least amount of content switch. 

Requesting you to allow to keep it. Thanks.

Comment (vinay0x) [reply app/javascript/src/apis/registrations.js:9]: Few things that immediately come to my mind:
1) For the IntelliSense to work, user needs to be aware of the name of the exported object right? They've to know the name to type `registrationsApi`. 
2) Number of times we reuse the same API file is very less.
3) If everything isn't configured perfectly, IntelliSense will use relative path for auto import.

Anyways if you insist on keeping this change, please make the change in other places.

Comment (yedhink) [reply app/javascript/src/apis/registrations.js:9]: I will make the changes everywhere.

- I thought we had a standard in place while writing Apis.
  * All apis reside in `apis/`
  * Naming of api file  = `pluralized-resource-name.js`. Eg:- `registrations.js`
  * Naming of the api object = `pluralizedResourceNameApi`. Eg:- `registrationsApi`
  * (Optional or Maximum try to) Try our best to name the actions within that api align with the actions associated with that resource. Eg:- `update, destroy, create etc`. 

If we don't have such a standard, then I suggest we can enforce it. Such a convention would help with easier debugging and easier coding when someone new starts with it. They only have to keep the `routes.rb` in mind. Rest they could just type without actually going into the `apis` folder and manually checking which file and object name should be used.

Comment (yedhink) [reply app/javascript/src/apis/registrations.js:9]: Will take this up in https://github.com/bigbinary/wheel/issues/453

Comment (vinay0x) [reply app/javascript/src/apis/registrations.js:9]: Handle in this PR itself. It's a small change. The APIs in this repos follows it. @yedhink _a

**Comment by yedhink** [comment]
@vinay0x _a

Have made the requested changes. 

Also fixes #453 . Have kept the resource name as `authentication` itself. But has been camelCased.

Let me know if we need to make anymore changes.

**Comment by vinay0x** [comment]
LGTM!
